src/mpid/ch4/generic/am/mpidig_am_msg.h

```diff
@@ -195,29 +195,22 @@ MPL_STATIC_INLINE_PREFIX int MPIDIG_recv_copy_seg(void *payload, MPI_Aint payloa
     MPIDIG_rreq_async_t *p = &(MPIDIG_REQUEST(rreq, req->async));
 
     if (p->recv_type == MPIDIG_RECV_DATATYPE) {
-        MPI_Aint max_data_size;
-        MPIR_Datatype_get_size_macro(MPIDIG_REQUEST(rreq, datatype), max_data_size);
-        max_data_size *= MPIDIG_REQUEST(rreq, count);
-
         MPI_Aint actual_unpack_bytes;
         MPIR_Typerep_unpack(payload, payload_sz,
                             MPIDIG_REQUEST(rreq, buffer),
                             MPIDIG_REQUEST(rreq, count),
                             MPIDIG_REQUEST(rreq, datatype), p->offset, &actual_unpack_bytes);
-        p->in_data_sz -= actual_unpack_bytes;
-        p->offset += actual_unpack_bytes;
-        if (payload_sz + p->offset - actual_unpack_bytes > max_data_size) {
-            /* did not fit */
-            rreq->status.MPI_ERROR = MPIDIG_ERR_TRUNCATE(p->offset, p->offset + p->in_data_sz);
-            MPIR_STATUS_SET_COUNT(rreq->status, p->offset);
-            return 1;
-        } else if (payload_sz > actual_unpack_bytes) {
+        if (!rreq->status.MPI_ERROR && payload_sz > actual_unpack_bytes) {
             /* basic element size mismatch */
             rreq->status.MPI_ERROR =
                 MPIR_Err_create_code(MPI_SUCCESS, MPIR_ERR_RECOVERABLE, __func__, __LINE__,
                                      MPI_ERR_TYPE, "**dtypemismatch", 0);
-            return 1;
-        } else if (p->in_data_sz == 0) {
+        }
+
+        p->in_data_sz -= payload_sz;
+        p->offset += payload_sz;
```

Comment:
```suggestion
        p->offset += actual_unpack_bytes;
```

Reply:
This is actually a main point of this PR. The `actual_unpack_bytes` can be less than the `payload_sz` (size of the segment), so adding `actual_unpack_bytes` may result  in  the next segment offset mis-aligned. We can't simply abort either, because we can't leave the remaining segments hanging.

Reply:
We need to make sure we receive all the data, but the buffer contents are going to be garbage. We should just make sure we don't go out of bounds.

Reply:
At this branch, it is already an error case -- we have a hole at the end of this segment. The focus is to receive the rest of the segments. If we adjust the offset by `actual_unpack_bytes`, then the next segments will be copied at wrong offsets -- all shifted by the size of "hole". I guess it probably doesn't matter match since we already claimed this is an error and user shouldn't care about data anyway. I was thinking at least the reset of the segments can be copied correctly so minimizing the data corruption or at least reduce the amount data noise. 

Why do you think adjust by `payload_sz` would result in out-of-bound ?

Reply:
My assumption is that if Typerep leaves a hole at the end of the unpack operation, then a future unpack starting at the end of the hole might not have enough space. This hinges on `in_data_sz` being equal to `dt_size` * `count`. But since this is the error case, it's probably not worth debating too vigorously. We could also run this through the AddressSanitizer to see if it flags anything.

Reply:
`in_data_sz` and `dt_size` is being checked up-front: https://github.com/pmodels/mpich/blob/0462b595ab25dd0495b6882e7058995b21bef417/src/mpid/ch4/generic/am/mpidig_am_msg.h#L29-L35

I concede that it probably doesn't matter how we adjust `offset` in this case. Since it doesn't matter, I prefer the uniformity this version provides:
```
       p->in_data_sz -= payload_sz;
       p->offset += payload_sz;
```
-- otherwise, it leads readers to ponder on the difference between `payload_sz` and `actual_unpack_bytes` when they doesn't matter, thus more difficult to read.

Let me rebase and run the tests.

Reply:
Note that, since we always use power of 2 as segment size (and headers observe 8-byte alignment), I don't think this error case (due to basic datatype mis-align) will ever arise. But since we added this error case, I think we'd better make it correct.

Reply:
Also, https://github.com/pmodels/mpich/blob/0462b595ab25dd0495b6882e7058995b21bef417/src/mpid/ch4/generic/am/mpidig_am_msg.h#L199-L202
`MPIR_Typerep_unpack` is supposed to check recv datatype type and size, so even if we erroneously give an out-of-bound offset, it should not actually write out-of-bound, right?

Reply:
Dataloop code doesn't appear to take offset into it's error determination. The yaksa Typerep impl doesn't look to have the same error checks, but they also might be down in the library itself. We also aren't capturing errors from the Typerep APIs...

Reply:
Both dataloop and yaksa have to seek to the offset before unpacking, right? I am assuming it has to bail out if the offset is beyond the datatype. Either that, or it will simply only take the remaining data, which will become 0 or negative for offset beyond datatype boundary, right?

Reply:
Yes, but Typereps are repeatable, e.g. vector. There is no guarantee dataloop or yaksa are tracking the bounds of the (buffer,type,count). In fact, they might explicitly not do that for performance purposes.

Reply:
> Yes, but Typereps are repeatable, e.g. vector. There is no guarantee dataloop or yaksa are tracking the bounds of the (buffer,type,count). In fact, they might explicitly not do that for performance purposes.

It has to calculate how many bytes beyond `offset` to copy, right? In the case of `offset` beyond bounds, the number of bytes will be negative or truncated to 0, right?

Reply:
Anyway, that was my assumption. It worth checking out and confirm for sure. If my assumption doesn't hold -- typerep just blindly copy -- then I am afraid there are more places to place guard.

Reply:
> > Yes, but Typereps are repeatable, e.g. vector. There is no guarantee dataloop or yaksa are tracking the bounds of the (buffer,type,count). In fact, they might explicitly not do that for performance purposes.
> 
> It has to calculate how many bytes beyond `offset` to copy, right? In the case of `offset` beyond bounds, the number of bytes will be negative or truncated to 0, right?

Copy size is passed in explicitly to Typerep_unpack. It is not modified by the offset.